### ISSUE TITLE
cabal-install doesn't respect ^C

### ISSUE THREAD

**Issue by phadej** [haskell/cabal#6322]
**Describe the bug**

One have to ^C a lot to make cabal cancel. This is super annoying

**To Reproduce**

`cabal v2-build` something with plenty (fresh to build) dependencies. Try to ^C build right away. It won't. (on Linux).

**Expected behavior**

`cabal` should cancel promptly.

**System information**

- Operating System: Linux
- cabal-install-3.0.0.0

**Additional context**

E.g. when built through `make`, make seems to forward ^C, and then dies itself, leaving "downloading" cabal on the background. That's bad.

Feels like some missing `bracket` somewhere.

**Comment by phadej** [comment]
```
^Cghc-pkg: interrupted
gi^Cghc-pkg: interrupted
^Cghc-pkg: interrupted
^Cghc-pkg: interrupted
^Cghc-pkg: interrupted
^C^C^Cghc-pkg: interrupted
```

**Comment by 23Skidoo** [comment]
I remember that we already had some code/tickets related to this, but don't remember where. @dcoutts?

**Comment by robx** [comment]
I might have a fix for this in #7929, but it's a bit unclear if there's a different bug. I see the described behaviour, but only when there are (multiple) packages to be downloaded, and I try to interrupt during download. I've managed to reproduce that bug, and fix it in the PR. From the description and terminal log here it's unclear whether downloading was involved.

**Comment by ulysses4ever** [comment]
@robx I think we should close it, do you agree? I just tried to Ctrl-C at various stages while building `lens` on Linux. It worked like a charm.

**Comment by robx** [comment]
Yes, that sounds good to me -- this report is Linux-specific, and to the best of my knowledge the issue is fixed there. (I think the fix didn't make it into 3.8, but should be there in current master and 3.10.)

I'll close, of course feel free to reopen this if it resurfaces, or if we should wait for the 3.10 release before closing the issue.

**Comment by ulysses4ever** [comment]
I was trying on 3.8 by the way. And  #7921 definitely got in 3.8.

**Comment by robx** [comment]
> I was trying on 3.8 by the way. And #7921 definitely got in 3.8.

I think it's #7929 that fixed this, which did make it into 3.8.

(The primary effect of #7921 is to ensure that `kill`ing `cabal run` kills whatever `cabal run` is running as well as `cabal` itself.)
